notebooks/Bayesian Changepoint Detection.py

```diff
@@ -0,0 +1,234 @@
+    def __init__(self, hazard, distribution):
+        self.hazard = hazard
+        self.distribution = distribution
+        self.time = 0
+    
+        self.Pr = np.ones(1)
+    
+    def step(self, x):
+        self.time += 1
+
+        H = self.hazard(np.arange(self.time))
+        old = self.Pr
+        self.Pr = np.zeros(self.time + 1)   # +1 because 0 <= r <= time
+        
+        predprob = self.distribution.pdf(x)
+        self.distribution.update(x)
+        
+        self.Pr[1:] = old * predprob * (1 - H)
+        self.Pr[0] = np.sum(old * predprob * H)
+        self.Pr /= np.sum(self.Pr)
+        
+        return self.Pr
+
+
+# In[33]:
+
+class Distribution(object):
+    
+    def __init__(self):
+        self.kappa = 0
+        self.mu = 0
+        
+    def pdf(self, data):
+        return 1
+    
+    def update(self, data):
+        kappa_update = self.kappa + 1.
+        mu_update = (self.kappa * self.mu + data) / (self.kappa + 1)
+        
+        self.kappa = kappa_update
+        self.mu = mu_update
+
+
+# In[34]:
+
+class Gaussian(Distribution):
+    """ Adapted from
+        http://engineering.richrelevance.com/
+            bayesian-analysis-of-normal-distributions-with-python/,
+        http://www.cs.ubc.ca/~murphyk/Papers/bayesGauss.pdf
+    """
+    
+    def __init__(self, kappa, mu, alpha, beta):
+        # Certainty about mean
+        self.kappa = kappa
+        # Estimation of Mean
+        self.mu = mu
+        # Estimation of Variance
+        # Paramater for Gamma Distribution
+        self.alpha = alpha
+        # Sum of Squares
+        self.beta = beta
+    
+    def pdf(self, data):
+        return stats.norm.pdf(x = data,
+                              loc = self.mu,
+                              scale = np.sqrt(self.beta / self.kappa))
+    
+    def update(self, data):
+        # Update Confidence
+        kappa_update = self.kappa + 1.
+        # Update Mean
+        mu_update = (self.kappa * self.mu + data) / (self.kappa + 1)
+        # Update Variance
+        alpha_update = self.alpha + .5
+        beta_update = self.beta + (self.kappa * (data - self.mu)**2) / (2. * (self.kappa + 1.))
+        
+        self.kappa = kappa_update
+        self.mu = mu_update
+        self.alpha = alpha_update
+        self.beta = beta_update
+
+
+# In[35]:
+
+class StudentT(Distribution):
+    """ Taken from https://github.com/hildensia/bayesian_changepoint_detection
+        Adapted to remove concatenation and standardize style
+    """
+    def __init__(self, kappa, mu, alpha, beta):
+        self.kappa = kappa
+        self.mu = mu
+        self.alpha = alpha
+        self.beta = beta
+
+    def pdf(self, data):
+        return stats.t.pdf(x=data, 
+                           df=2*self.alpha,
+                           loc=self.mu,
+                           scale=np.sqrt(self.beta * (self.kappa+1) / (self.alpha *
+                               self.kappa)))
+
+    def update(self, data):
```

Comment:
So, does this still work? Because if we're at time `t`, we have `t+1` different distributions, one for each possible value of `r`. It seems like now, we only have one distribution for everything.

I like this architecture better (transparent state), but I'm pretty sure we need to edit the `BayesianChangepointDetection` logic too.